cmd/api-errors.go

```diff
@@ -22,6 +22,11 @@ import (
 	"fmt"
 	"net/http"
 
+	"github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
+	"google.golang.org/api/googleapi"
+
+	minio "github.com/minio/minio-go"
```

Comment:
@harshavardhana instead of importing gcs/azure packages/oss in the cmd package, this is cleaner:

```
func toAPIError(ctx context.Context, err error) APIError {
	if err == nil {
		return noError
	}
	if gateayErr, ok := err.(GatewayError); ok {
		return APIError{
			gatewayErr.Code,
			gatewayErr.Description,
			gatewayErr.HTTPStatusCode,
		}
	}
	return errorCodes.ToAPIErr(toAPIErrorCode(ctx, err))
}
```
```
type GatewayError struct {
	Code           string
	Description    string
	HTTPStatusCode int
}
```
and
```
--- a/cmd/gateway/gcs/gateway-gcs.go
+++ b/cmd/gateway/gcs/gateway-gcs.go
@@ -325,7 +325,11 @@ func gcsToObjectError(err error, params ...string) error {
                }
                err = minio.BucketNotEmpty{Bucket: bucket}
        default:
-               err = fmt.Errorf("Unsupported error reason: %s", reason)
+               err = minio.GatewayError{
+                       Code:           reason,
+                       Description:    message,
+                       HTTPStatusCode: googleAPIErr.Code,
+               }
        }
 
        return err
```

Reply:
There are limitations in this approach, it loses information which I would like being accessible for future use. This is also the point of avoiding custom translation of errors from underneath in gateway. respective SDKs already do this for us. 

We don't need to comply with a common data structure as it may not be the case in all scenarios and its overkill to keep translating from one error to another form of error. We already vendorize all these libraries with their entire source code - I don't know what is the hesitation of using them - its one place and one function - which shows clearly what we are doing with each of such requirement.

I am not able to see a convincing reason to add GatewayError.